provider/openstack/firewaller.go

```diff
@@ -70,16 +71,341 @@ type firewallerFactory struct {
 
 // GetFirewaller implements FirewallerFactory
 func (f *firewallerFactory) GetFirewaller(env environs.Environ) Firewaller {
-	return &defaultFirewaller{environ: env.(*Environ)}
+	return &switchingFirewaller{env: env.(*Environ)}
+}
+
+type switchingFirewaller struct {
+	env *Environ
+
+	mu sync.Mutex
+	fw Firewaller
+}
+
+func (f *switchingFirewaller) initFirewaller() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.fw != nil {
+		return nil
+	}
+
+	client := f.env.client()
+	if !client.IsAuthenticated() {
+		if err := authenticateClient(client); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	base := firewallerBase{environ: f.env}
+	if f.env.supportsNeutron() {
+		f.fw = &neutronFirewaller{base}
+	} else {
+		f.fw = &legacyNovaFirewaller{base}
+	}
+	return nil
+}
+
+func (f *switchingFirewaller) OpenPorts(ports []network.PortRange) error {
+	if err := f.initFirewaller(); err != nil {
+		return errors.Trace(err)
+	}
+	return f.fw.OpenPorts(ports)
+}
+
+func (f *switchingFirewaller) ClosePorts(ports []network.PortRange) error {
+	if err := f.initFirewaller(); err != nil {
```

Comment:
Why does this have to happen in each call? Can things be rearranged that there's a constructor which ensures the setup in initFirewaller happens then? 

Or is there a requirement/benefit for doing the authentication lazily?